core/src/main/scala/org/ensime/core/javac/JavaCompletion.scala

```diff
@@ -208,21 +208,26 @@ trait JavaCompletion extends Helpers with SLF4JLogging {
     importing: Boolean,
     caseSense: Boolean
   ): List[CompletionInfo] = {
-    val candidates = typeElement(info, target).map { el =>
-      el match {
-        case tel: TypeElement => {
-          val elements: Elements = info.getElements()
-          elements.getAllMembers(tel).flatMap { e =>
-            filterElement(info, e, prefix, caseSense, importing, false)
-          }
+
+    typeElement(info, target).map {
+
+      case tel: TypeElement =>
+
+        def isAccessible(el: Element): Boolean = {
+          val path = info.getTrees.getPath(info.getCompilationUnit, target)
+          val scope = info.getTrees.getScope(path)
+          info.getTrees.isAccessible(scope, el, info.getTypes.getDeclaredType(tel))
         }
-        case e => {
-          log.warn("Unrecognized type element " + e)
-          List()
+
+        info.getElements.getAllMembers(tel).filter(isAccessible).flatMap { el =>
+          filterElement(info, el, prefix, caseSense, importing, false)
         }
-      }
-    }.getOrElse(List())
-    candidates.toList
+
+      case e =>
+        log.warn("Unrecognized type element " + e)
+        List.empty
+
+    }.map(_.toList).getOrElse(List.empty)
```

Comment:
do you know about `breakOut`? Highly recommended watching: https://skillsmatter.com/skillscasts/6939-optimising-scala-for-fun-and-profit (you'll never call `.toList`, `.toMap` or `.toSet` again...)